spark/v3.4/spark/src/main/java/org/apache/iceberg/spark/actions/RewritePositionDeletesSparkAction.java

```diff
@@ -0,0 +1,501 @@
+    }
+
+    List<PositionDeleteGroupRewriteResult> rewriteResults =
+        commitResults.stream()
+            .map(RewritePositionDeleteGroup::asResult)
+            .collect(Collectors.toList());
+    return ImmutableResult.builder().rewriteResults(rewriteResults).build();
+  }
+
+  Stream<RewritePositionDeleteGroup> toGroupStream(
+      RewriteExecutionContext ctx,
+      Map<StructLike, List<List<PositionDeletesScanTask>>> groupsByPartition) {
+    Stream<RewritePositionDeleteGroup> rewriteFileGroupStream =
+        groupsByPartition.entrySet().stream()
+            .flatMap(
+                e -> {
+                  StructLike partition = e.getKey();
+                  List<List<PositionDeletesScanTask>> scanGroups = e.getValue();
+                  return scanGroups.stream()
+                      .map(
+                          tasks -> {
+                            int globalIndex = ctx.currentGlobalIndex();
+                            int partitionIndex = ctx.currentPartitionIndex(partition);
+                            // as coerced partition used for map, return original partition
+                            PositionDeleteGroupInfo info =
+                                ImmutablePositionDeleteGroupInfo.builder()
+                                    .globalIndex(globalIndex)
+                                    .partitionIndex(partitionIndex)
+                                    .partition(partition)
+                                    .build();
+                            return new RewritePositionDeleteGroup(info, tasks);
+                          });
+                });
+
+    return rewriteFileGroupStream.sorted(rewriteGroupComparator());
+  }
+
+  private Comparator<RewritePositionDeleteGroup> rewriteGroupComparator() {
+    switch (rewriteJobOrder) {
+      case BYTES_ASC:
+        return Comparator.comparing(RewritePositionDeleteGroup::rewrittenBytes);
+      case BYTES_DESC:
+        return Comparator.comparing(
+            RewritePositionDeleteGroup::rewrittenBytes, Comparator.reverseOrder());
+      case FILES_ASC:
+        return Comparator.comparing(RewritePositionDeleteGroup::numRewrittenDeleteFiles);
+      case FILES_DESC:
+        return Comparator.comparing(
+            RewritePositionDeleteGroup::numRewrittenDeleteFiles, Comparator.reverseOrder());
+      default:
+        return (fileGroupOne, fileGroupTwo) -> 0;
+    }
+  }
+
+  void validateAndInitOptions() {
+    Set<String> validOptions = Sets.newHashSet(rewriter.validOptions());
+    validOptions.addAll(VALID_OPTIONS);
+
+    Set<String> invalidKeys = Sets.newHashSet(options().keySet());
+    invalidKeys.removeAll(validOptions);
+
+    Preconditions.checkArgument(
+        invalidKeys.isEmpty(),
+        "Cannot use options %s, they are not supported by the action or the rewriter %s",
+        invalidKeys,
+        rewriter.description());
+
+    rewriter.init(options());
+
+    maxConcurrentFileGroupRewrites =
```

Comment:
optional: Sometimes, we include `this.XXX` to indicate we assign a value to an instance variable. Up to you.